plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java

```diff
@@ -391,96 +449,112 @@ public void eventOccurred(RequestCompletedEvent eventArg) {
         }
     }
 
-    /**
-     * Building our own input stream instead of using the SDK's {@link com.microsoft.azure.storage.blob.BlobInputStream}
-     * because that stream is highly inefficient in both memory and CPU use.
-     */
-    private static class BlobInputStream extends InputStream {
+    private static class AzureInputStream extends StorageInputStream {
+        private final BlobAsyncClient client;
+        private final ByteBuffer buffer;
+        private final int maxRetries;
+        private final long firstReadOffset;
+        private final int firstReadLength;
+
+        private AzureInputStream(final BlobAsyncClient client,
+                                 long rangeOffset,
+                                 Long rangeLength,
+                                 int chunkSize,
+                                 long contentLength,
+                                 int maxRetries) throws IOException {
+            super(rangeOffset, rangeLength, chunkSize, contentLength);
+            this.client = client;
+            this.buffer = ByteBuffer.allocate(chunkSize);
+            this.maxRetries = maxRetries;
+
+            // Read eagerly the first chunk so we can throw early if the
+            // blob doesn't exist
+            this.firstReadOffset = rangeOffset;
+            this.firstReadLength = (int) Math.min(chunkSize, contentLength - rangeOffset);
+            executeRead(firstReadLength, rangeOffset);
+        }
+
+        @Override
+        protected ByteBuffer dispatchRead(int readLength, long offset) throws IOException {
+            // If the request is for the first chunk, don't download it again
+            // Since we disabled marking in this InputStream, the offset only advances
+            // and never goes back requesting the firstReadOffset again
+            if (offset != firstReadOffset || readLength != firstReadLength) {
+                executeRead(readLength, offset);
```

Comment:
Sorry for not pointing this out earlier, but I'm a little worried about this stream.
Is it actually faster than what we currently have with the old SDK? We still have the same inefficient behavior here as far as I can tell. We fetch data into a 4M buffer lazily and copy it into the smaller read buffer that comes from other snapshot functionality like we did before, just using non-blocking IO instead of blocking IO in the background.
This is in contrast to the other SDKs where we just execute one request for the full file that we then read in a blocking fashion (thereby avoiding the latency of a request every 4M and getting much better performance).
Can't we build this much more efficiently by starting a `Flux` on the whole stream and then blockingly iterating through it via `toStream()` to get similar behavior to a full blocking read triggered by a single request?